lib/measures/tbd/measure.rb

```diff
@@ -18,32 +18,118 @@
   require_relative "resources/log.rb"
 end
 
-# TBD exit strategy when faced with warnings/errors. The generated file targets
-# a design context ( >= WARN ) ... change TBD log_level for debugging purposes.
-# By default, log_status is set below DEBUG while log_level is set @WARN.
-def exitTBD(runner, out_dir, io)
+##
+# TBD exit strategy. Outputs TBD model content/results if out && io are TRUE.
+# Generates log errors and warnings, even if io or out are FALSE.
+#
+# @param [Runner] runner OpenStudio Measure runner
+# @param [Bool] out True if user wishes to output detailed TBD content/results
+# @param [Hash] io TBD input/output content
+# @param [Hash] surfaces TBD derated surfaces
+#
+# @return [Bool] Returns true if TBD Measure is successful.
+def exitTBD(runner, out = false, io = nil, surfaces = nil)
+  # Generated files target a design context ( >= WARN ) ... change TBD log_level
+  # for debugging purposes. By default, log_status is set below DEBUG while
+  # log_level is set @WARN.
+  #
   # TBD.set_log_level(TBD::DEBUG)
-  unless TBD.logs.empty? || TBD.status < TBD.log_level
-    msgs = []
-    TBD.logs.each do |l|
-      msg = "#{l[:time]} (#{TBD.tag(l[:level])}) #{l[:msg]}"
-      runner.registerError(msg) if l[:level] == TBD::FATAL
-      runner.registerWarning(msg) unless l[:level] == TBD::FATAL
-      msgs << msg
+
+  status = TBD.msg(TBD.status)
+  status = TBD.msg(TBD::INFO) if TBD.status.zero?
+
+  unless io && surfaces
+    status = "Halting all TBD processes - running OpenStudio" unless TBD.fatal?
+  end
+
+  io = {} unless io
+  description = "Thermal Bridging and Derating (TBD) - see github.com/rd2/tbd"
+  io[:description] = description unless io.has_key?(:description)
+
+  unless io.has_key?(:schema)
+    io[:schema] = "https://github.com/rd2/tbd/blob/master/tbd.schema.json"
+  end
+
+  if surfaces
+    surfaces.each do |id, surface|
+      next if TBD.fatal?
+      next unless surface.has_key?(:ratio)
+      ratio  = format "%3.1f", surface[:ratio]
+      name   = id.rjust(15, " ")
+      output = "#{name} RSi derated by #{ratio}%"
+      TBD.log(TBD::INFO, output)
+      runner.registerInfo(output)
     end
-    runner.registerWarning("Consult 'tbd.log' - under /files") unless TBD.fatal?
-
-    out_path = File.join(out_dir, "tbd.log")
-    File.open(out_path, "w") do |l|
-      l.write msgs.join("\n")
-      begin
-        l.fsync
-      rescue StandardError
-        l.flush
-      end
+  end
+
+  tbd_log = { date: Time.now, status: status }
+
+  tbd_msgs = []
+  TBD.logs.each do |l|
+    tbd_msgs << { level: TBD.tag(l[:level]), message: l[:message] }
+    if l[:level] > TBD::INFO
+      runner.registerWarning("(#{TBD.tag(l[:level])}) #{l[:message]}")
+    else
+      runner.registerInfo("(#{TBD.tag(l[:level])}) #{l[:message]}")
```

Comment:
Early in _exitTBD()_, FATAL errors are registered as _warnings_. TBD & OpenStudio are nonetheless halted at the very end of _exitTBD_ (if FATAL).